### ISSUE TITLE
Generic NamedTuples and TypedDict

### ISSUE THREAD

**Issue by crusaderky** [python/mypy#7719]
mypy 0.730 doesn't work correctly with classes that descend both from NamedTuple or TypedDict and from Generic. dataclasses instead work fine.

```python
from dataclasses import dataclass
from typing import Generic, NamedTuple, TypeVar
from typing_extensions import TypedDict


T = TypeVar("T")

@dataclass
class A(Generic[T]):
    x: T


class B(NamedTuple, Generic[T]):
    x: T


class C(TypedDict, Generic[T]):
    x: T


"foo" + A(x=1).x
"foo" + B(x=1).x
"foo" + C(x=1)["x"]
```
mypy output:
```
21: error: Unsupported operand types for + ("str" and "int")
22: error: Unsupported operand types for + ("str" and "T")
22: error: Argument "x" to "B" has incompatible type "int"; expected "T"
23: error: Unsupported operand types for + ("str" and "T")
23: error: Incompatible types (expression has type "int", TypedDict item "x" has type "T")
```
Expected output:
```
21: error: Unsupported operand types for + ("str" and "int")
22: error: Unsupported operand types for + ("str" and "int")
23: error: Unsupported operand types for + ("str" and "int")
```

**Comment by ilevkivskyi** [comment]
Generic named tuples and typed dicts are not supported, see feature-request issues:
* Generic named tuples: https://github.com/python/mypy/issues/685
* Generic typed dicts: https://github.com/python/mypy/issues/3863

The bug currently is that these are not flagged as errors, see https://github.com/python/mypy/issues/6200
